libqpdf/QPDFJob.cc

```diff
@@ -2350,30 +2350,34 @@ QPDFJob::handleTransformations(QPDF& pdf)
         for (auto& ph: dh.getAllPages()) {
             ++pageno;
             QPDFObjectHandle page = ph.getObjectHandle();
-            for (auto& iter2: ph.getImages()) {
-                std::string name = iter2.first;
-                QPDFObjectHandle& image = iter2.second;
-                ImageOptimizer* io = new ImageOptimizer(
-                    *this,
-                    m->oi_min_width,
-                    m->oi_min_height,
-                    m->oi_min_area,
-                    image);
-                std::shared_ptr<QPDFObjectHandle::StreamDataProvider> sdp(io);
-                if (io->evaluate(
-                        "image " + name + " on page " +
-                        std::to_string(pageno))) {
-                    QPDFObjectHandle new_image = pdf.newStream();
-                    new_image.replaceDict(image.getDict().shallowCopy());
-                    new_image.replaceStreamData(
-                        sdp,
-                        QPDFObjectHandle::newName("/DCTDecode"),
-                        QPDFObjectHandle::newNull());
-                    ph.getAttribute("/Resources", true)
-                        .getKey("/XObject")
-                        .replaceKey(name, new_image);
-                }
-            }
+            QPDFPageObjectHelper(page).forEachImage(
+                true,
+                [&, pageno, &pdf = pdf](
+                    QPDFObjectHandle& obj,
+                    QPDFObjectHandle& xobj_dict,
+                    std::string const& key) mutable {
+                    std::string name = key;
+                    QPDFObjectHandle& image = obj;
+                    ImageOptimizer* io = new ImageOptimizer(
+                        *this,
+                        m->oi_min_width,
+                        m->oi_min_height,
+                        m->oi_min_area,
+                        image);
+                    std::shared_ptr<QPDFObjectHandle::StreamDataProvider> sdp(
+                        io);
+                    if (io->evaluate(
+                            "image " + name + " on page " +
+                            std::to_string(pageno))) {
+                        QPDFObjectHandle new_image = pdf.newStream();
+                        new_image.replaceDict(image.getDict().shallowCopy());
+                        new_image.replaceStreamData(
+                            sdp,
```

Comment:
... and
```cpp
                                std::move(io),
```
it here.

Reply:
Feel free to push to my branch. I'll need to spend some time to understand this.